File: go/arrow/ipc/writer_test.go
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package ipc

import (
	"bytes"
	"encoding/binary"
	"fmt"
	"math"
	"strings"
	"testing"

	"github.com/apache/arrow/go/v13/arrow"
	"github.com/apache/arrow/go/v13/arrow/array"
	"github.com/apache/arrow/go/v13/arrow/bitutil"
	"github.com/apache/arrow/go/v13/arrow/internal/flatbuf"
	"github.com/apache/arrow/go/v13/arrow/memory"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

// reproducer from ARROW-13529
func TestSliceAndWrite(t *testing.T) {
	alloc := memory.NewGoAllocator()
	schema := arrow.NewSchema([]arrow.Field{
		{Name: "s", Type: arrow.BinaryTypes.String},
	}, nil)

	b := array.NewRecordBuilder(alloc, schema)
	defer b.Release()

	b.Field(0).(*array.StringBuilder).AppendValues([]string{"foo", "bar", "baz"}, nil)
	rec := b.NewRecord()
	defer rec.Release()

	sliceAndWrite := func(rec arrow.Record, schema *arrow.Schema) {
		slice := rec.NewSlice(1, 2)
		defer slice.Release()

		fmt.Println(slice.Columns()[0].(*array.String).Value(0))

		var buf bytes.Buffer
		w := NewWriter(&buf, WithSchema(schema))
		w.Write(slice)
		w.Close()
	}

	assert.NotPanics(t, func() {
		for i := 0; i < 2; i++ {
			sliceAndWrite(rec, schema)
		}
	})
}

func TestNewTruncatedBitmap(t *testing.T) {
	alloc := memory.NewCheckedAllocator(memory.DefaultAllocator)
	defer alloc.AssertSize(t, 0)

	assert.Nil(t, newTruncatedBitmap(alloc, 0, 0, nil), "input bitmap is null")

	buf := memory.NewBufferBytes(make([]byte, bitutil.BytesForBits(8)))
	defer buf.Release()

	bitutil.SetBit(buf.Bytes(), 0)
	bitutil.SetBit(buf.Bytes(), 2)
	bitutil.SetBit(buf.Bytes(), 4)
	bitutil.SetBit(buf.Bytes(), 6)

	assert.Same(t, buf, newTruncatedBitmap(alloc, 0, 8, buf), "no truncation necessary")

	result := newTruncatedBitmap(alloc, 1, 7, buf)
	defer result.Release()
	for i, exp := range []bool{false, true, false, true, false, true, false} {
		assert.Equal(t, exp, bitutil.BitIsSet(result.Bytes(), i), "truncate for offset")
	}

	buf = memory.NewBufferBytes(make([]byte, 128))
	defer buf.Release()
	bitutil.SetBitsTo(buf.Bytes(), 0, 128*8, true)

	result = newTruncatedBitmap(alloc, 0, 8, buf)
	defer result.Release()
	assert.Equal(t, 64, result.Len(), "truncate to smaller buffer")
	assert.Equal(t, 8, bitutil.CountSetBits(result.Bytes(), 0, 8))
}

func TestGetZeroBasedValueOffsets(t *testing.T) {
	alloc := memory.NewCheckedAllocator(memory.DefaultAllocator)
	defer alloc.AssertSize(t, 0)

	vals := []string{"a", "b", "c", "d", "e", "f", "g", "h", "i", "j"}
	b := array.NewStringBuilder(alloc)
	defer b.Release()
	b.AppendValues(vals, nil)

	arr := b.NewArray()
	defer arr.Release()

	env := &recordEncoder{mem: alloc}

	offsets, err := env.getZeroBasedValueOffsets(arr)
	require.NoError(t, err)
	defer offsets.Release()
	assert.Equal(t, 44, offsets.Len(), "include all offsets if array is not sliced")

	sl := array.NewSlice(arr, 0, 4)
	defer sl.Release()

	offsets, err = env.getZeroBasedValueOffsets(sl)
	require.NoError(t, err)
	defer offsets.Release()
	assert.Equal(t, 20, offsets.Len(), "trim trailing offsets after slice")
}

func TestWriterCatchPanic(t *testing.T) {
	alloc := memory.NewGoAllocator()
	schema := arrow.NewSchema([]arrow.Field{
		{Name: "s", Type: arrow.BinaryTypes.String},
	}, nil)

	b := array.NewRecordBuilder(alloc, schema)
	defer b.Release()

	b.Field(0).(*array.StringBuilder).AppendValues([]string{"foo", "bar", "baz"}, nil)
	rec := b.NewRecord()
	defer rec.Release()

	// mess up the first offset for the string column
	offsetBuf := rec.Column(0).Data().Buffers()[1]
	bitutil.SetBitsTo(offsetBuf.Bytes(), 0, 32, true)

	buf := new(bytes.Buffer)

	writer := NewWriter(buf, WithSchema(schema))
	assert.EqualError(t, writer.Write(rec), "arrow/ipc: unknown error while writing: runtime error: slice bounds out of range [-1:]")
}

func TestWriterMemCompression(t *testing.T) {
	mem := memory.NewCheckedAllocator(memory.NewGoAllocator())
	defer mem.AssertSize(t, 0)

	schema := arrow.NewSchema([]arrow.Field{
		{Name: "s", Type: arrow.BinaryTypes.String},
	}, nil)

	b := array.NewRecordBuilder(mem, schema)
	defer b.Release()

	b.Field(0).(*array.StringBuilder).AppendValues([]string{"foo", "bar", "baz"}, nil)
	rec := b.NewRecord()
	defer rec.Release()

	var buf bytes.Buffer
	w := NewWriter(&buf, WithAllocator(mem), WithSchema(schema), WithZstd())
	defer w.Close()

	require.NoError(t, w.Write(rec))
}

func TestWriteWithCompressionAndMinSavings(t *testing.T) {
	mem := memory.NewCheckedAllocator(memory.DefaultAllocator)
	defer mem.AssertSize(t, 0)

	// a small batch that is known to be compressible
	batch, _, err := array.RecordFromJSON(mem, arrow.NewSchema([]arrow.Field{
		{Name: "n", Type: arrow.PrimitiveTypes.Int64, Nullable: true}}, nil),
		strings.NewReader(`[
			{"n": 0}, {"n": 1}, {"n": 2}, {"n": 3}, {"n": 4},
			{"n": 5}, {"n": 6}, {"n": 7}, {"n": 8}, {"n": 9}]`))
	require.NoError(t, err)
	defer batch.Release()

	prefixedSize := func(buf *memory.Buffer) int64 {
		if buf.Len() < arrow.Int64SizeBytes {
			return 0
		}
		return int64(binary.LittleEndian.Uint64(buf.Bytes()))
	}
	contentSize := func(buf *memory.Buffer) int64 {
		return int64(buf.Len()) - int64(arrow.Int64SizeBytes)
	}

	for _, codec := range []flatbuf.CompressionType{flatbuf.CompressionTypeLZ4_FRAME, flatbuf.CompressionTypeZSTD} {
		enc := newRecordEncoder(mem, 0, 5, true, codec, 1, nil)
		var payload Payload
		require.NoError(t, enc.encode(&payload, batch))
		assert.Len(t, payload.body, 2)

		// compute the savings when body buffers are compressed unconditionally.
		// We also validate that our test batch is indeed compressible.
		uncompressedSize, compressedSize := prefixedSize(payload.body[1]), contentSize(payload.body[1])
		assert.Less(t, compressedSize, uncompressedSize)
		assert.Greater(t, compressedSize, int64(0))
		expectedSavings := 1.0 - float64(compressedSize)/float64(uncompressedSize)

		compressEncoder := newRecordEncoder(mem, 0, 5, true, codec, 1, &expectedSavings)
		payload.Release()
		payload.body = payload.body[:0]
		require.NoError(t, compressEncoder.encode(&payload, batch))
		assert.Len(t, payload.body, 2)
		assert.Equal(t, uncompressedSize, prefixedSize(payload.body[1]))
		assert.Equal(t, compressedSize, contentSize(payload.body[1]))

		payload.Release()
		payload.body = payload.body[:0]
		// slightly bump the threshold. the body buffer should now be prefixed
		// with -1 and its content left uncompressed
		minSavings := math.Nextafter(expectedSavings, 1.0)
		compressEncoder.minSpaceSavings = &minSavings
		require.NoError(t, compressEncoder.encode(&payload, batch))
		assert.Len(t, payload.body, 2)
		assert.EqualValues(t, -1, prefixedSize(payload.body[1]))
		assert.Equal(t, uncompressedSize, contentSize(payload.body[1]))
		payload.Release()
		payload.body = payload.body[:0]

		for _, outOfRange := range []float64{math.Nextafter(1.0, 2.0), math.Nextafter(0, -1)} {
			compressEncoder.minSpaceSavings = &outOfRange
			err := compressEncoder.encode(&payload, batch)
			assert.ErrorIs(t, err, arrow.ErrInvalid)
			assert.ErrorContains(t, err, "minSpaceSavings not in range [0,1]")
		}
	}
}

File: cpp/src/arrow/json/writer_test.cc
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

#include "gtest/gtest.h"

#include <memory>
#include <string>
#include <vector>

#include "arrow/buffer.h"
#include "arrow/io/memory.h"
#include "arrow/ipc/writer.h"
#include "arrow/json/writer.h"
#include "arrow/record_batch.h"
#include "arrow/result.h"
#include "arrow/testing/gtest_util.h"
#include "arrow/testing/matchers.h"
#include "arrow/type.h"
#include "arrow/type_fwd.h"

namespace arrow {
namespace json {

class TestWriteJSON : public ::testing::Test {
 protected:
  template <typename Data>
  Result<std::string> ToJSONString(const Data& data, const WriteOptions& options) {
    std::shared_ptr<io::BufferOutputStream> out;
    ARROW_ASSIGN_OR_RAISE(out, io::BufferOutputStream::Create());

    RETURN_NOT_OK(WriteJSON(data, options, out.get()));
    ARROW_ASSIGN_OR_RAISE(std::shared_ptr<Buffer> buffer, out->Finish());
    return std::string(reinterpret_cast<const char*>(buffer->data()), buffer->size());
  }

  Result<std::string> ToJSONStringUsingWriter(const Table& data,
                                              const WriteOptions& options) {
    std::shared_ptr<io::BufferOutputStream> out;
    ARROW_ASSIGN_OR_RAISE(out, io::BufferOutputStream::Create());
    ARROW_ASSIGN_OR_RAISE(auto writer, MakeJSONWriter(out, data.schema(), options));
    TableBatchReader reader(data);
    reader.set_chunksize(1);
    std::shared_ptr<RecordBatch> batch;
    RETURN_NOT_OK(reader.ReadNext(&batch));
    while (batch != nullptr) {
      RETURN_NOT_OK(writer->WriteRecordBatch(*batch));
      RETURN_NOT_OK(reader.ReadNext(&batch));
    }
    RETURN_NOT_OK(writer->Close());
    ARROW_ASSIGN_OR_RAISE(std::shared_ptr<Buffer> buffer, out->Finish());
    return std::string(reinterpret_cast<const char*>(buffer->data()), buffer->size());
  }
};

TEST_F(TestWriteJSON, EmptyBatch) {
  auto schema = ::arrow::schema({field("a", int64()), field("b", utf8())});
  auto batch = RecordBatchFromJSON(schema, "[]");

  WriteOptions options;
  ASSERT_OK_AND_ASSIGN(std::string json, ToJSONString(*batch, options));
  EXPECT_EQ(json, "");
}

TEST_F(TestWriteJSON, SimpleTypes) {
  auto schema = ::arrow::schema({
      field("int64", int64()),
      field("double", float64()),
      field("string", utf8()),
      field("bool", boolean()),
      field("date32", date32()),
      field("timestamp", timestamp(TimeUnit::SECOND)),
  });

  auto batch = RecordBatchFromJSON(schema, R"([
    {"int64": 42, "double": 3.14, "string": "hello", "bool": true, "date32": 172800000, "timestamp": 172800000},
    {"int64": -1, "double": 2.71, "string": "world", "bool": false, "date32": 259200000, "timestamp": 259200000}
  ])");

  WriteOptions options;
  ASSERT_OK_AND_ASSIGN(std::string json, ToJSONString(*batch, options));
  EXPECT_EQ(
      json,
      R"({"int64":42,"double":3.14,"string":"hello","bool":true,"date32":172800000,"timestamp":172800000}
{"int64":-1,"double":2.71,"string":"world","bool":false,"date32":259200000,"timestamp":259200000}
)");
}

TEST_F(TestWriteJSON, NullValues) {
  auto schema = ::arrow::schema({
      field("a", int64()),
      field("b", utf8()),
      field("c", boolean()),
      field("d", null()),
  });

  auto batch = RecordBatchFromJSON(schema, R"([
    {"a": 1, "b": "x", "c": true, "d": null},
    {"a": null, "b": "y", "c": null, "d": null},
    {"a": 3, "b": "z", "c": false, "d": null}
  ])");

  WriteOptions options;
  options.emit_null = true;
  ASSERT_OK_AND_ASSIGN(std::string json, ToJSONString(*batch, options));
  EXPECT_EQ(json,
            R"({"a":1,"b":"x","c":true,"d":null}
{"a":null,"b":"y","c":null,"d":null}
{"a":3,"b":"z","c":false,"d":null}
)");
}

TEST_F(TestWriteJSON, SkipNullValues) {
  auto schema = ::arrow::schema({
      field("a", int64()),
      field("b", utf8()),
      field("c", boolean()),
      field("d", null()),
  });

  auto batch = RecordBatchFromJSON(schema, R"([
    {"a": 1, "b": "x", "c": true, "d": null},
    {"a": null, "b": "y", "c": null, "d": null},
    {"a": 3, "b": null, "c": false, "d": null}
  ])");

  WriteOptions options;
  options.emit_null = false;
  ASSERT_OK_AND_ASSIGN(std::string json, ToJSONString(*batch, options));
  EXPECT_EQ(json,
            R"({"a":1,"b":"x","c":true}
{"b":"y"}
{"a":3,"c":false}
)");
}

TEST_F(TestWriteJSON, NestedStruct) {
  auto schema = ::arrow::schema({
      field("id", int64()),
      field("data", struct_({
                        field("name", utf8()),
                        field("value", int32()),
                    })),
  });

  auto batch = RecordBatchFromJSON(schema, R"([
    {"id": 1, "data": {"name": "foo", "value": 42}},
    {"id": 2, "data": {"name": "bar", "value": 100}}
  ])");

  WriteOptions options;
  ASSERT_OK_AND_ASSIGN(std::string json, ToJSONString(*batch, options));
  EXPECT_EQ(json,
            R"({"id":1,"data":{"name":"foo","value":42}}
{"id":2,"data":{"name":"bar","value":100}}
)");
}

TEST_F(TestWriteJSON, ListType) {
  auto schema = ::arrow::schema({
      field("id", int64()),
      field("values", list(int32())),
  });

  auto batch = RecordBatchFromJSON(schema, R"([
    {"id": 1, "values": [1, 2, 3]},
    {"id": 2, "values": [4, 5]},
    {"id": 3, "values": []}
  ])");

  WriteOptions options;
  ASSERT_OK_AND_ASSIGN(std::string json, ToJSONString(*batch, options));
  EXPECT_EQ(json,
            R"({"id":1,"values":[1,2,3]}
{"id":2,"values":[4,5]}
{"id":3,"values":[]}
)");
}

TEST_F(TestWriteJSON, NestedListAndStruct) {
  auto schema = ::arrow::schema({
      field("id", int64()),
      field("items", list(struct_({
                         field("key", utf8()),
                         field("value", int64()),
                     }))),
  });

  auto batch = RecordBatchFromJSON(schema, R"([
    {"id": 1, "items": [{"key": "a", "value": 1}, {"key": "b", "value": 2}]},
    {"id": 2, "items": []}
  ])");

  WriteOptions options;
  ASSERT_OK_AND_ASSIGN(std::string json, ToJSONString(*batch, options));
  EXPECT_EQ(json,
            R"({"id":1,"items":[{"key":"a","value":1},{"key":"b","value":2}]}
{"id":2,"items":[]}
)");
}

TEST_F(TestWriteJSON, FromTable) {
  auto schema = ::arrow::schema({
      field("a", int64()),
      field("b", utf8()),
  });

  auto batch = RecordBatchFromJSON(schema, R"([
    {"a": 1, "b": "x"},
    {"a": 2, "b": "y"}
  ])");

  ASSERT_OK_AND_ASSIGN(auto table, Table::FromRecordBatches({batch}));

  WriteOptions options;
  ASSERT_OK_AND_ASSIGN(std::string json, ToJSONString(*table, options));
  EXPECT_EQ(json,
            R"({"a":1,"b":"x"}
{"a":2,"b":"y"}
)");
}

TEST_F(TestWriteJSON, FromRecordBatchReader) {
  auto schema = ::arrow::schema({
      field("a", int64()),
      field("b", utf8()),
  });

  auto batch1 = RecordBatchFromJSON(schema, R"([
    {"a": 1, "b": "x"}
  ])");
  auto batch2 = RecordBatchFromJSON(schema, R"([
    {"a": 2, "b": "y"}
  ])");

  ASSERT_OK_AND_ASSIGN(auto reader, RecordBatchReader::Make({batch1, batch2}));

  WriteOptions options;
  ASSERT_OK_AND_ASSIGN(std::string json, ToJSONString(reader, options));
  EXPECT_EQ(json,
            R"({"a":1,"b":"x"}
{"a":2,"b":"y"}
)");
}

TEST_F(TestWriteJSON, BatchSize) {
  auto schema = ::arrow::schema({
      field("a", int64()),
  });

  auto batch = RecordBatchFromJSON(schema, R"([
    {"a": 1}, {"a": 2}, {"a": 3}, {"a": 4}, {"a": 5}
  ])");

  WriteOptions options;
  options.batch_size = 2;  // Process 2 rows at a time
  ASSERT_OK_AND_ASSIGN(std::string json, ToJSONString(*batch, options));
  EXPECT_EQ(json,
            R"({"a":1}
{"a":2}
{"a":3}
{"a":4}
{"a":5}
)");
}

TEST_F(TestWriteJSON, UsingWriter) {
  auto schema = ::arrow::schema({
      field("a", int64()),
      field("b", utf8()),
  });

  auto batch = RecordBatchFromJSON(schema, R"([
    {"a": 1, "b": "x"},
    {"a": 2, "b": "y"}
  ])");

  ASSERT_OK_AND_ASSIGN(auto table, Table::FromRecordBatches({batch}));

  WriteOptions options;
  ASSERT_OK_AND_ASSIGN(std::string json, ToJSONStringUsingWriter(*table, options));
  EXPECT_EQ(json,
            R"({"a":1,"b":"x"}
{"a":2,"b":"y"}
)");
}

TEST_F(TestWriteJSON, LargeString) {
  auto schema = ::arrow::schema({
      field("a", large_utf8()),
  });

  auto batch = RecordBatchFromJSON(schema, R"([
    {"a": "hello"},
    {"a": "world"}
  ])");

  WriteOptions options;
  ASSERT_OK_AND_ASSIGN(std::string json, ToJSONString(*batch, options));
  EXPECT_EQ(json,
            R"({"a":"hello"}
{"a":"world"}
)");
}

TEST_F(TestWriteJSON, FixedSizeList) {
  auto schema = ::arrow::schema({
      field("a", fixed_size_list(int32(), 3)),
  });

  auto batch = RecordBatchFromJSON(schema, R"([
    {"a": [1, 2, 3]},
    {"a": [4, 5, 6]}
  ])");

  WriteOptions options;
  ASSERT_OK_AND_ASSIGN(std::string json, ToJSONString(*batch, options));
  EXPECT_EQ(json,
            R"({"a":[1,2,3]}
{"a":[4,5,6]}
)");
}

TEST_F(TestWriteJSON, OptionsValidation) {
  WriteOptions options;
  options.batch_size = 0;
  EXPECT_RAISES_WITH_MESSAGE_THAT(
      Invalid, ::testing::HasSubstr("batch_size must be at least 1"), options.Validate());

  options.batch_size = 1024;
  ASSERT_OK(options.Validate());
}

}  // namespace json
}  // namespace arrow
